session4/autocomplete.js

```diff
@@ -0,0 +1,81 @@
+class TrieNode {
+    constructor(data) {
+        this.data = data;
+        this.children = new Map();
+        this.isTerminal = false;
+    }
+
+    getChild(data) {
+        return this.children.get(data);
+    }
+}
+
+class Trie {
+    constructor() {
+        this.root = new TrieNode(' ');
+    }
+
+    addWord(word) {
+        let current = this.root;
+        
+        for(let i = 0; i < word.length; i++) {
+            let letter = word[i];
+            let child = current.getChild(letter);
+            if(child !== undefined) {
+                current = child;
+            }
+            else {
+                current.children.set(letter ,new TrieNode(letter));
+                current = current.getChild(letter);
+            }
+        }
+        current.isTerminal = true;
+    }
+
+    build(list) {
+        list.forEach(word => {
+            this.addWord(word);
+        });
+    }
+
+    getNext(prefix) {
+        let lastNode = this.getLast(prefix);
+        return [...lastNode.children.keys()];
+    }
+
+    getLast(prefix) {
+        let lastNode = this.root;
+        for(let i = 0; i < prefix.length; i++) {
+            let next = lastNode.children.get(prefix[i]);
+            if(!next) {
+                break;
+            }
+            lastNode = next;
+        }
+
+        return lastNode;
+    }
+
+    getAutoComplete(prefix) {
+        let lastNode = this.getLast(prefix);
+        let output = [];
+        lastNode.children.forEach(child => {
```

Comment:
Esto solamente te da los hijos directos que sean terminales. Sin embargo esto tiene que regresar todas las palabras completas donde prefix sea un prefijo, por ejemplo, si tu diccionario tienen la palabra: "casa", getAutoComplete("c") debe de regresar "casa".

Reply:
Si lo noté pero ayer no se me ocurría como resolverlo 😅 
Acabo de mandar un fix, lo único que se me ocurrió fue volver a llamar el mismo `getAutoComplete` pero con el prefix + el next string

corrí las pruebas con `getAutocomplete('c')`  y si me regresa todas las pablabras que empiezan con c. Pero no se si haya una manera distinta de lograrlo
```
[ 'cat', 'car', 'cap', 'can' ]
```

Reply:
Es correcto, all llamar recursivamente estas haciendo un traversal, lo cual es correcto.